### DISCUSSION TITLE
[wasm] huge difference in performance

### DISCUSSION THREAD

**Discussion by dkonecny-oracle** [zxing-cpp/zxing-cpp#511] [Q&A] [answered]
Im following your example https://zxing-cpp.github.io/zxing-cpp/zxing_viddemo.html and in Chrome on my MacBook Pro the performance of that example is pretty good. If I open console I can see there:
```
zxing_viddemo.html:94 [Violation] 'requestAnimationFrame' handler took 239ms
zxing_viddemo.html:94 [Violation] 'requestAnimationFrame' handler took 169ms
zxing_viddemo.html:94 [Violation] 'requestAnimationFrame' handler took 208ms
zxing_viddemo.html:94 [Violation] 'requestAnimationFrame' handler took 230ms
```
When I run almost the same code in my own app using my own build of zxing_reader.js/wasm the performance gets even better:
```
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 65ms
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 66ms
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 75ms
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 59ms
```
BUT when I try to use zxing_reader.js/wasm from your latest release (ie https://github.com/zxing-cpp/zxing-cpp/releases/download/v2.0.0/wasm-artifacts.zip) then the same app get sluggish and console shows:
```
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 2073ms
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 1295ms
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 1474ms
main-start-page.js:98 [Violation] 'requestAnimationFrame' handler took 1627ms
```
Any explanation for that? Id prefer to use your released binaries. I tested also your 1.4 release and that one seems to have good performance (just like your live demo). For building I just did:
```
emcmake cmake ../wrappers/wasm
cmake --build .
```
using cmake 3.25.2 on Mac 12.6.2 (using Chrome 109).
Thx,
-David

**Comment by axxel** [comment]
Thanks for bringing this to my attention. I'm not using the WASM wrapper myself and based on feedback I get here it seems like not many people do either. My very first thought was that maybe the CI build accidentally got switched to build in Debug mode, but that does not seem to be the case.

Regarding your observations:
1. the binary files on github.io are horrendously outdated. That means they are definitively less capable and quite possibly less performant.
2. did you get those results by building current HEAD or 2.0 on your machine?
3. I have absolutely no idea why those 2.0 wasm-artifacts files would be so slow. According to the CI build log I don't see anything suspicious:
```
Run emcmake cmake -Swrappers/wasm -Bbuild
  emcmake cmake -Swrappers/wasm -Bbuild
  shell: /usr/bin/bash -e {0}
  env:
    BUILD_TYPE: Release
    PATH: /home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master:/home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master/upstream/emscripten:/home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master/node/14.15.5_64bit/bin:/home/runner/.local/bin:/opt/pipx_bin:/home/runner/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    EMSDK: /home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master
    EM_CONFIG: /home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master/.emscripten
    EMSDK_NODE: /home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master/node/14.15.5_64bit/bin/node
configure: cmake -Swrappers/wasm -Bbuild -DCMAKE_TOOLCHAIN_FILE=/home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR="/home/runner/work/_temp/c94816d3-22a4-4755-b2e9-7e3caca0d006/emsdk-master/node/14.15.5_64bit/bin/node"
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Performing Test COMPILER_NEEDS_FLOAT_STORE
-- Performing Test COMPILER_NEEDS_FLOAT_STORE - Failed
-- Configuring done
-- Generating done
```

[... a few minutes later ...]
I tried to get a feel for this myself now and experimented a bit with the live-demo on my machine. I don't know how to get those timing measurements (in either Firefox or Chromium) but my first run was indeed _very_ slow. Then I switched my build from `MinSizeRel` to `Release` and suddenly it was a lot faster. I played a bit with the "Performance" tab in the Chromium Inspector and it looks like the `readBarcodeFromPixmap` function executes in under 10ms on my 3 year old Dell XPS.

Since the above log from the CI build explicitly states that a `Release` build was setup, I still don't understand... hm.

**Comment by dkonecny-oracle** [comment]
Thanks for prompt evaluation and looking into it! This does not block me right now and was just FYI.

Chrome automatically logs "[Violation] 'requestAnimationFrame' handler took" message when time spent in the callback (to process camera frame) is longer than 16ms (I read that on the net) and many users complained that the log should have INFO log level or be removed. But it came handy for me to report some perf measurements back to you as the time corresponds to time spent in "readBarcodeFromPixmap".

I pulled master branch (did not think about it!) and build it.

No idea what might be causing this. Let me know if I could help you in any way.
-David

**Reply by axxel** [reply]
Thanks for clarifying the origin of the "Violation" Warning. Now I understand why I did not see it. But then again, why is the current HEAD revision built as `Release` almost 10x faster on my machine compared to yours? Maybe you could build it with `RelWithDebInfo`. That allows for some more insights from the Performance tab of the Chrome(ium) browser:
![image](https://user-images.githubusercontent.com/1038346/217412483-01f57f9d-efe9-4cba-930a-2e1646e6f4e0.png)

Maybe we get a clue where your build spends the extra time compared to my local one. No need to build the 2.0 tag...

EDIT: what could easily explain the difference is, if you used a higher video resolution than I do (640x480)...

**Comment by dkonecny-oracle** [comment]
Different video resolution sounded like a good explanation but it does not make big difference. My video stretches with page width so  1680x720 or 450x250 lead to very similar timings. Im not sure how to pass RelWithDebInfo into the build. Neither Im very skilled in chrome performance diagnostics. But I run 2.0 (left in the screenshot) and compared it to my build (right one) and the list of method calls is very different. Probably does not help.
-David
<img width="1041" alt="image" src="https://user-images.githubusercontent.com/48036491/217658542-137cd863-556a-41a1-9558-28cc37b13cbe.png">

**Comment by axxel** [comment, marked as answer]
Further analysis showed that compiling with `MinSizeRel` was indeed the cause for the slow performance. I added some measurements to the [README](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/wasm/README.mdl). What I failed to see earlier is that with no explicit build mode specified, the CMakeLists.txt forced the `MinSizeRel` build.

If you can confirm that the latest [build artifacts](https://github.com/zxing-cpp/zxing-cpp/suites/10875443233/artifacts/548704579) are solving your issue, I'll probably simply upload those files to the 2.0 release.

**Reply by dkonecny-oracle** [reply]
The latest built artefacts have great performance. So it makes sense to me to republish them as 2.0.

The only crinkle is that I have not specified any build mode (I dont even know how to do that) and my build was run with "-Os" and it had good performance. ;-) 

-David

**Reply by axxel** [reply]
Thanks for your confirmation. Your "-Os" observation is indeed seriously questioning my analysis. :-/. But what is also still pretty unclear is why your best performance is still ~6-8x slower than what I observe here. I can hardly believe that your MPB should indeed be so much slower than my 3 year old Dell. When you say "great performance", what does that mean in numbers? Could you maybe run the downloaded build artifacts with the new [demo_cam_reader](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/wasm/demo_cam_reader.html) that I just added and let me know what your console says about the 'requestAnimationFrame' handler?

**Reply by axxel** [reply]
Update: after upgrading my emsdk version to the latest, the "-Os" build is substantially faster than before (14ms vs 300ms) but still only half as fast as the "-O3" build. This now explains everything except the difference between your runtime measurements and mine (65ms vs 14ms).

**Reply by dkonecny-oracle** [reply]
I really appreciate the time you are giving this problem and your desire to nail it down and resolve all remaining open questions.

I took your demo_cam_reader and added some simple diagnostics (as Chrome does not log requestAnimationFrame violation for short times), ie:
```
	var time =  Date.now();
	var buffer = zxing._malloc(sourceBuffer.byteLength);
	zxing.HEAPU8.set(sourceBuffer, buffer);
	var result = zxing.readBarcodeFromPixmap(buffer, imgWidth, imgHeight, mode, format);
	zxing._free(buffer);
	var time2 =  Date.now();
	totalTime += (time2-time);
	totalCount++;
	console.log(">>> zxing.routines average.time="+(Math.round(totalTime/totalCount))+" current.time="+(time2-time));
```
and created simple website with multiple copies of this file with different versions of zxing binaries. Im using "Normal" mode (not sure what is the difference from the Fast mode part from being much faster - ie what do I loose by using it?) and I can see these **average times**:
* your fixed binaries: 24ms
* my build binaries with -Os option: 26ms
* my build binaries with -O3 option: 28ms
* broken 2.0 release: 112ms

Which is good conclusion I think - the figures are all similar and <30ms is great performance.

I cannot anymore reproduce ultra slow times I reported initially. I can also confirm that changing video size to 1600x800 does increase average time to 56ms (still very reasonable!)
I think we can rest this case? wdyt?
Thx,
-David

**Reply by axxel** [reply]
> not sure what is the difference from the Fast mode part from being much faster - ie what do I loose by using it?)

"Fast" means disabling the `tryHarder`, `tryInvert`, `tryDownscale` and `tryRotate` flags. See the `DecodeHints.h` for details.

> and I can see these **average times**:

That looks pretty reasonable, although your findings with even faster -Os than -O3 builds are in line with what I found. But that might be explained by a different emsdk version or maybe just by the different platforms involved.


> I think we can rest this case? wdyt? Thx, -David

Agrred.
